crates/matrix-sdk/src/latest_events/mod.rs

```diff
@@ -0,0 +1,861 @@
+    collections::{HashMap, HashSet},
+    ops::{ControlFlow, Not},
+    sync::Arc,
+};
+
+use futures_util::{select, FutureExt};
+pub use latest_event::LatestEvent;
+use matrix_sdk_common::executor::{spawn, AbortOnDrop, JoinHandleExt as _};
+use ruma::{EventId, OwnedEventId, OwnedRoomId, RoomId};
+use tokio::sync::{broadcast, mpsc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use tracing::error;
+
+use crate::{
+    event_cache::{EventCache, RoomEventCacheGenericUpdate},
+    send_queue::SendQueue,
+};
+
+/// The entry point to fetch the [`LatestEvent`] for rooms or threads.
+#[derive(Clone, Debug)]
+pub struct LatestEvents {
+    state: Arc<LatestEventsState>,
+}
+
+/// The state of [`LatestEvents`].
+#[derive(Debug)]
+struct LatestEventsState {
+    /// All the registered rooms, i.e. rooms the latest events are computed for.
+    registered_rooms: Arc<RegisteredRooms>,
+
+    /// The task handle of the
+    /// [`listen_to_event_cache_and_send_queue_updates_task`].
+    _listen_task_handle: AbortOnDrop<()>,
+
+    /// The task handle of the [`compute_latest_events_task`].
+    _computation_task_handle: AbortOnDrop<()>,
+}
+
+impl LatestEvents {
+    /// Create a new [`LatestEvents`].
+    pub(crate) fn new(event_cache: EventCache, send_queue: SendQueue) -> Self {
+        let (room_registration_sender, room_registration_receiver) = mpsc::channel(32);
+        let (latest_event_queue_sender, latest_event_queue_receiver) = mpsc::unbounded_channel();
+
+        let registered_rooms = Arc::new(RegisteredRooms::new(room_registration_sender));
+
+        // The task listening to the event cache and the send queue updates.
+        let listen_task_handle = spawn(listen_to_event_cache_and_send_queue_updates_task(
+            registered_rooms.clone(),
+            room_registration_receiver,
+            event_cache,
+            send_queue,
+            latest_event_queue_sender,
+        ))
+        .abort_on_drop();
+
+        // The task computing the new latest events.
+        let computation_task_handle = spawn(compute_latest_events_task(
+            registered_rooms.clone(),
+            latest_event_queue_receiver,
+        ))
+        .abort_on_drop();
+
+        Self {
+            state: Arc::new(LatestEventsState {
+                registered_rooms,
+                _listen_task_handle: listen_task_handle,
+                _computation_task_handle: computation_task_handle,
+            }),
+        }
+    }
+
+    /// Start listening to updates (if not already) for a particular room, and
+    /// fetch the [`LatestEvent`] for this room.
+    ///
+    /// It returns `None` if the room doesn't exist.
+    ///
+    /// Note that a read lock guard is returned if the [`LatestEvent`] exists.
+    /// One doesn't want to keep this lock alive for too long, otherwise the
+    /// whole [`LatestEvents`] API is blocked. Use [`LatestEvent::get`] or
+    /// [`LatestEvent::subscribe`] as quick as possible, and then drop this
+    /// lock guard.
```

Comment:
This sounds a bit footgun-y, do we really need to expose the internal lock guard or can we maybe instead expose the underlying `get/value` and `subscribe` methods directly (similar to how `forget_thread` just makes magic happen)? e.g. subscribe & get room + subscribe & get thread